Require macOS 10.13 and add support for pause/resume

karaggeorge opened the pull request:
Closes #2 
Closes #5 
Closes #1

Also replaces #12 (re-implemented those changes here)

So, this has a bunch of changes:

- Adds swift-argument-parser for the CLI part to manage commands, since now they have increased a lot
- Adds a two way communication capability between the main aperture process and the node instance via:
  - Ability to send events (and get a response)
  - Ability to listen to events

Aperture sends events like `onStart`, `onPause` etc that we can listen to, and we can send events to Aperture like `pause`, `resume` to access functionality.

The communication happens using `DistributedNotificationCenter` and specifically mapped event names, inspired by how `electron-better-ipc` works.
This is sandboxed to a single aperture process using a randomly generated `processId` which is included in the events.

Also, the events can answer back using another unique id that is generated on the spot for each event. This way two parallel events of the same type won't get their responses confused.

I've tested the regular flow, including `pause`, `resume` and `isPaused` to ensure it works.

A weird observation. This code:
```
await recorder.startRecording(options);
console.log('Started recording');

recorder.isFileReady.then(() => console.log('File is ready'))

await recorder.pause();
console.log('Paused recording');

await delay(5000)

await recorder.resume();
console.log('Resume recording');
```

Outputs:
```
Started recording
Paused recording
Resumed recording
File is ready
```

And in the output the first frame is after the `resume` call (no time-skip). So, we could theoretically pause right after start, then play a countdown timer or do whatever and then resume, to ensure we start at a convenient time for us.

karaggeorge commented on Package.swift:
```diff
@@ -1,4 +1,4 @@
-// swift-tools-version:5.2
+// swift-tools-version:5.1
```
I'm on Mojave, so I can't run 5.2. Had to set to 5.1 for development. We can bump again before merging

karaggeorge commented on Package.swift:
```diff
@@ -15,13 +15,15 @@ let package = Package(
   dependencies: [
-    .package(url: "https://github.com/wulkano/Aperture", from: "0.2.0")
+    .package(url: "https://github.com/wulkano/Aperture", from: "0.2.0"),
+    .package(url: "https://github.com/apple/swift-argument-parser", from: "0.1.0")
```
Again, used `0.1.0` because I can only run on Swift 5.1, but it had everything I needed for this

karaggeorge replied:
After I'm done with the rest of the comments if you want to pull down and migrate, that'd work too. I don't know if any breaking changes happened. I just can't even fetch 0.2.0 without using Swift tools 5.2, so I couldn't test it

sindresorhus replied:
We can keep it 5.1 until you can upgrade.

karaggeorge commented:
@sindresorhus I'm sure my swift code needs cleanup, feel free to push any updates needed on the branch if you don't want to comment on everything 😄 

Also if you can test it locally as well that'd be great

karaggeorge commented:
I haven't added the 1 sec delay we discussed in the other PR. I didn't test it here. I'll test it locally in a bit, and update with the results

karaggeorge commented:
Looks like there's still that 1s delay. Both after just resolving `startRecording()` and if you start/pause/resume, there's 1s delay after `resume()` resolves. 
Running it a few times, I've gotten anywhere between 0.6 and 0.9s delay

However, from the moment before you call `.resume()` and it resolving it's about .1s

So, if we add the 1s delay on both promises, we could call `.resume()` and know we have about 1s to perform all animations before the recording starts

sindresorhus commented:
> So, if we add the 1s delay on both promises, we could call .resume() and know we have about 1s to perform all animations before the recording starts

👍

karaggeorge commented:
Added the 1s delay to both the resume and startRecording calls

sindresorhus commented on index.js:
```diff
@@ -43,7 +44,7 @@ class Aperture {
   } = {}) {
     this.processId = getRandomId();
-    return new Promise((resolve, reject) => {
+    return new Promise(async (resolve, reject) => {
```
It's an anti-pattern to use `async` in `new Promise` as if the async function rejects, it causes an unhandled rejection. It's better then to use an async-IIFE and try/catch inside it.

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+    answer(nil)
+  }
+
+  func answer(_ data: Any?) {
```
```suggestion
  func answer(_ data: Any? = nil) {
```

and you can drop `func answer()`

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+    var payload = [AnyHashable: Any]()
+
+    if data != nil {
+      payload["data"] = "\(data!)"
```
Don't force-unwrap. Use guard or if-let. Applies in other places too.

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+  data: Any?,
+  using callback: @escaping (ApertureNotification) -> Void
+) {
+  let responseName = "aperture.\(processId).\(event).response.\(UUID().uuidString)"
```
`responseIdentifier` would be more correct, I think.

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  DistributedNotificationCenter.default().postNotificationName(
+    .Name("aperture.\(processId).\(event)"),
```
`"aperture.\(processId).\(event)` is used in three places. I think it could benefit from being moved into a function.

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+  sendEvent(processId: processId, event: event, data: nil) { _ in }
+}
+
+func sendEvent(processId: String, event: String, data: Any?) {
```
```suggestion
func sendEvent(processId: String, event: String, data: Any? = nil) {
```

and you can drop the above method.

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+    payload["data"] = "\(data!)"
+  }
+
+  var observer: Any?
```
```suggestion
  var observer: AnyObject?
```

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+  }
+}
+
+func sendEvent(
```
Can you namespace these methods? (enum with static members)

sindresorhus commented on Sources/ApertureCLI/notifications.swift:
```diff
@@ -0,0 +1,109 @@
+  )
+}
+
+func sendEvent(processId: String, event: String, using callback: @escaping (ApertureNotification) -> Void) {
```
`processId` should be a `Int`.

karaggeorge replied:
Looks like we default it to "main"
Should it be a required arg? Or should it default to some kind of random Int?

sindresorhus commented:
Bump

karaggeorge commented:
Will be working on this next, when I have some downtime

sindresorhus commented:
@karaggeorge Bump :)

YoucefHQ commented:
+1, this would be amazing to have for users. Thank you 🙏🏽

karaggeorge commented:
Sorry for all the delay 😅 
@sindresorhus I updated the dependency versions (since I'm on 10.15 now 🎉 ) and updated based on your comments.

I also updated the js code a bit to match the PR that went in since this was opened, so now the startRecording resolves with nothing, but isFileReady resolves with the file path.

Also, bumped the minimum version to 10.13 since new Aperture is also 10.13

Because of the above two, we probably need a major version bump when publishing.

I'll also work on adding some typings in a separate PR when I get some time

YoucefHQ commented:
@karaggeorge Thank you so much for doing this. I'm a Product Manager, and so many of our users asked us for the ability to pause/stop their recordings. We were able to build it on Windows, but we were waiting for Aperture-Node to support it. So thank you on behalf of our users. 🙏🏽

karaggeorge commented:
@YoucefHQ no problem. If you don't mind me asking, what are you using as a windows alternative for aperture-node?

YoucefHQ commented:
@karaggeorge We a native library on Windows called ScreenRecorderLib https://github.com/sskodje/ScreenRecorderLib
